core/src/Cuda/Kokkos_Cuda_Error.hpp

```diff
@@ -63,17 +63,45 @@ void cuda_device_synchronize(const std::string& name);
 void cuda_stream_synchronize(const cudaStream_t stream,
                              const std::string& name);
 
-void cuda_internal_error_throw(cudaError e, const char* name,
-                               const char* file = nullptr, const int line = 0);
+[[noreturn]] void cuda_internal_error_throw(cudaError e, const char* name,
+                                            const char* file = nullptr,
+                                            const int line   = 0);
+
+[[noreturn]] void cuda_internal_error_abort(cudaError e, const char* name,
+                                            const char* file = nullptr,
+                                            const int line   = 0);
+
+#ifdef KOKKOS_ENABLE_CXX17
+#define KOKKOS_IMPL_FALLTHROUGH [[fallthrough]];
+#else
+#define KOKKOS_IMPL_FALLTHROUGH
+#endif
 
 inline void cuda_internal_safe_call(cudaError e, const char* name,
                                     const char* file = nullptr,
                                     const int line   = 0) {
-  if (cudaSuccess != e) {
-    cuda_internal_error_throw(e, name, file, line);
+  // 1. Success -> normal continuation.
+  // 2. Error codes for which, to continue using CUDA, the process must be
+  //    terminated and relaunched -> call abort on the host-side.
+  // 3. Any other error code -> throw a runtime error.
+  switch (e) {
+    case cudaSuccess: break;
+    case cudaErrorIllegalAddress: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorAssert: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorHardwareStackError: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorIllegalInstruction: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorMisalignedAddress: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorInvalidAddressSpace: KOKKOS_IMPL_FALLTHROUGH
+    case cudaErrorInvalidPc: KOKKOS_IMPL_FALLTHROUGH
```

Comment:
Do you really get warnings if the body of each `case` is empty or is that for stylistic reasons?

Reply:
Just anticipating them

Reply:
If we don't need them (which coincides with my prior experience), I prefer not defining `KOKKOS_IMPL_FALLTHROUGH` and removing it here.

Reply:
Sure will remove